File: ValidPalindrome.py
#https://leetcode.com/problems/valid-palindrome
class Solution:
    def isPalindrome(self, s: str) -> bool:
        x=''
        for i in s:
            if i.isalnum():
                x+=i.lower()
        return x==x[::-1]

# Solution 2
#with helper function check ascii value
class Solution:
    def isPalindrome(self, s: str) -> bool:
        l,r=0,len(s)-1
        while l<r:
            while l<r and not self.alphaNum(s[l]):
                l+=1
            while r>l and not self.alphaNum(s[r]):
                r-=1
            if s[l].lower()!=s[r].lower():
                return False
            l+=1
            r-=1    
        return True
    def alphaNum(self,x):
        return (ord("a")<= ord(x) <= ord("z") or
        ord("A")<= ord(x) <= ord("Z") or
        ord("0")<= ord(x) <= ord("9"))
        

File: ValidPalindrome.java
//https://leetcode.com/problems/valid-palindrome

// Solution 1
public class ValidPalindrome {
    public boolean isPalindrome(String s) {
        String x="";
        String rev="";
        for(int i=0;i<s.length();i++){
            if (Character.isLetterOrDigit(s.charAt(i))){
                x+=Character.toLowerCase(s.charAt(i));
            }
        }
        for(int i=x.length()-1;i>=0;i--){
            rev+=x.charAt(i);
        }
        if(x.equals(rev)){
            return true;
        }
        else{
            return false;
        }
    }
}

//Solution 2 Use StringBuilder
class Solution {
    public boolean isPalindrome(String s) {
        StringBuilder x = new StringBuilder();
        for(int i =0;i<s.length();i++){
            char c = s.charAt(i);
            if(Character.isLetterOrDigit(c)){
                x.append(Character.toLowerCase(c));
            }
        }
        return x.toString().equals(x.reverse().toString());
    }
}